### PR TITLE
feat: add interchain token service encoding/decoding api

### PR DESCRIPTION
**Pull request by milapsheth**
## Description

[AXE-4071]

## Todos

- [x] Unit tests
- [ ] Manual tests
- [ ] Documentation
- [ ] Connect epics/issues

## Other Notes

[AXE-4071]: https://axelarnetwork.atlassian.net/browse/AXE-4071?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/interchain-token-service-api/src/msg.rs [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+    // Returns a list of chains registered with the router
+    // The list is paginated by:
+    // - start_after: the chain name to start after, which the next page of results should start.
+    // - limit: limit the number of chains returned, default is u32::MAX.
+    #[returns(Vec<ChainEndpoint>)]
+    Chains {
+        start_after: Option<ChainName>,
+        limit: Option<u32>,
+    },
```
Comment (cjcobb23) [inline packages/interchain-token-service-api/src/msg.rs:20]: What's the use case for this? Who/what will query this that can't just query the router?

File: packages/interchain-token-service-api/src/primitives.rs [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+use cosmwasm_std::{HexBinary, Uint256};
+use error_stack::Report;
+
+#[cw_serde]
+#[derive(Eq)]
+pub struct TokenId {
+    #[serde(with = "axelar_wasm_std::hex")]
+    #[schemars(with = "String")]
+    pub id: [u8; 32],
+}
```
Comment (cjcobb23) [inline packages/interchain-token-service-api/src/primitives.rs:12]: consider just `pub type TokenId = [u8;32];`

File: packages/interchain-token-service-api/src/primitives.rs [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+#[cw_serde]
+#[derive(Eq)]
+pub enum ITSMessage {
+    InterchainTransfer {
+        token_id: TokenId,
+        source_address: HexBinary,
+        destination_address: HexBinary,
+        amount: Uint256,
+        data: HexBinary,
```
Comment (cjcobb23) [inline packages/interchain-token-service-api/src/primitives.rs:34]: Is this the full payload or the hash?

File: packages/interchain-token-service-api/src/primitives.rs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    Gateway = 5,
+}
+
+#[cw_serde]
+#[derive(Eq)]
+pub enum ITSMessage {
+    InterchainTransfer {
+        token_id: TokenId,
+        source_address: HexBinary,
+        destination_address: HexBinary,
```
Comment (cjcobb23) [inline packages/interchain-token-service-api/src/primitives.rs:32]: Should probably use the `Address` type that we use for `Message`. It's defined in `router_api::primitives`. Will make composability a bit cleaner.

File: packages/interchain-token-service-api/src/primitives.rs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        amount: Uint256,
+        data: HexBinary,
+    },
+    DeployInterchainToken {
+        token_id: TokenId,
+        name: String,
+        symbol: String,
+        decimals: u8,
+        minter: HexBinary,
```
Comment (cjcobb23) [inline packages/interchain-token-service-api/src/primitives.rs:41]: If this is an address, I'd prefer to use `Address` in `router_api::primitives`

File: packages/interchain-token-service-api/src/client.rs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+            contract_addr: self.address.to_string(),
+            msg: to_json_binary(&msg).expect("msg should always be serializable"),
+            funds: vec![],
+        }
+    }
+
+    pub fn route(&self, msgs: Vec<Message>) -> Option<WasmMsg> {
+        ignore_empty(msgs).map(|msgs| self.execute(&ExecuteMsg::RouteMessages(msgs)))
+    }
+}
```
Comment (cjcobb23) [inline packages/interchain-token-service-api/src/client.rs:22]: This looks like a client for the router. Why is this here?

File: packages/interchain-token-service-api/src/primitives.rs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+#[cw_serde]
+#[derive(Eq)]
+pub enum ITSMessage {
+    InterchainTransfer {
+        token_id: TokenId,
+        source_address: HexBinary,
+        destination_address: HexBinary,
+        amount: Uint256,
+        data: HexBinary,
+    },
```
Comment (cjcobb23) [inline packages/interchain-token-service-api/src/primitives.rs:35]: Why is there no source or destination chain here?

File: packages/interchain-token-service-api/src/abi.rs [hunk 8]
```diff
@@ -0,0 +1,427 @@
...
+    struct DeployTokenManager {
+        uint256 messageType;
+        bytes32 tokenId;
+        uint256 tokenManagerType;
+        bytes params;
+    }
+
+    struct RoutedCall {
+        uint256 messageType;
+        string chain;
```
Comment (cjcobb23) [inline packages/interchain-token-service-api/src/abi.rs:47]: Is this the destination chain or the source chain?

File: packages/interchain-token-service-api/src/msg.rs (hunk 1 above)
Comment (milapsheth) [reply packages/interchain-token-service-api/src/msg.rs:20]: ah, checked in by mistake

File: packages/interchain-token-service-api/src/primitives.rs (hunk 3 above)
Comment (milapsheth) [reply packages/interchain-token-service-api/src/primitives.rs:34]: Full payload. The whole struct gets encoded into the GMP payload that we approve over. But because ITS hub needs to receive the GMP payload to process, the ITS contract data also gets included. While we don't store this payload, it is going to affect p2p bandwidth and get included into the block data unfortunately. Not including this would require breaking changes on the relayers to support ITS GMP execute calls in a special way.

File: packages/interchain-token-service-api/src/primitives.rs (hunk 5 above)
Comment (milapsheth) [reply packages/interchain-token-service-api/src/primitives.rs:41]: same comment as above

File: packages/interchain-token-service-api/src/primitives.rs (hunk 7 above)
Comment (milapsheth) [reply packages/interchain-token-service-api/src/primitives.rs:35]: That info comes from the GMP Message struct. The true destination chain is included in `ITSRoutedMessage`, to be used for the next GMP call

File: packages/interchain-token-service-api/src/abi.rs (hunk 8 above)
Comment (milapsheth) [reply packages/interchain-token-service-api/src/abi.rs:47]: I'll change the name to remote_chain and add a comment. It's used for both. When ITS edge contract sends a RoutedCall, the remote_chain is the true destination, and when it receives the GMP call from ITS hub it's the true source_chain.

File: packages/interchain-token-service-api/src/client.rs (hunk 6 above)
Comment (milapsheth) [reply packages/interchain-token-service-api/src/client.rs:22]: same, got checked in by mistake

File: packages/interchain-token-service-api/src/primitives.rs (hunk 2 above)
Comment (milapsheth) [reply packages/interchain-token-service-api/src/primitives.rs:12]: since TokenId will be a public facing struct, we'd like to serialize it as hex. That's not doable if it's a type alias, right?

File: packages/interchain-token-service-api/src/abi.rs [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+        uint256 tokenManagerType;
+        bytes params;
+    }
+
+    struct RoutedCall {
+        uint256 messageType;
+        /// Remote chain name.
+        /// ITS edge source contract -> ITS Hub GMP call: Set to the true destination chain name.
+        /// ITS Hub -> ITS edge destination contract: Set to the true source chain name.
+        string remote_chain;
```
Comment (cjcobb23) [inline packages/interchain-token-service-api/src/abi.rs:55]: Is there a way we can not overload this field like this, and use some sort of enum? I find this confusing.

File: packages/interchain-token-service-api/src/primitives.rs [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+            0 => TokenManagerType::NativeInterchainToken,
+            1 => TokenManagerType::MintBurnFrom,
+            2 => TokenManagerType::LockUnlock,
+            3 => TokenManagerType::LockUnlockFee,
+            4 => TokenManagerType::MintBurn,
+            5 => TokenManagerType::Gateway,
+            _ => return Err(Report::new(Error::InvalidEnum)),
+        })
+    }
+}
```
Comment (cgorenflo) [inline packages/interchain-token-service-api/src/primitives.rs:87]: use `strum_macros::FromRepr` to generate this instead

File: packages/interchain-token-service-api/src/primitives.rs [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+#[cw_serde]
+#[derive(Eq, Copy)]
+#[repr(u8)]
+pub enum TokenManagerType {
+    NativeInterchainToken = 0,
+    MintBurnFrom = 1,
+    LockUnlock = 2,
+    LockUnlockFee = 3,
+    MintBurn = 4,
+    Gateway = 5,
```
Comment (cgorenflo) [inline packages/interchain-token-service-api/src/primitives.rs:24]: no need to number them all, rust will do the enumeration automatically

File: packages/interchain-token-service-api/src/primitives.rs [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    },
+}
+
+#[cw_serde]
+#[derive(Eq)]
+pub struct ITSRoutedMessage {
+    /// Remote chain name.
+    /// ITS edge source contract -> ITS Hub GMP call: Set to the true destination chain name.
+    /// ITS Hub -> ITS edge destination contract: Set to the true source chain name.
+    pub remote_chain: Address,
```
Comment (cgorenflo) [inline packages/interchain-token-service-api/src/primitives.rs:57]: as CJ mentioned, this is really weird behaviour

File: packages/interchain-token-service-api/src/abi.rs [hunk 13]
```diff
@@ -0,0 +1,469 @@
...
+        bytes message;
+    }
+}
+
+impl ITSRoutedMessage {
```
Comment (cgorenflo) [inline packages/interchain-token-service-api/src/abi.rs:60]: why is this defined in a separate module?

File: packages/interchain-token-service-api/src/abi.rs [hunk 14]
```diff
@@ -0,0 +1,469 @@
...
+            ITSMessage::DeployTokenManager {
+                token_id,
+                token_manager_type,
+                params,
+            } => DeployTokenManager {
+                messageType: MessageType::DeployTokenManager.into(),
+                tokenId: FixedBytes::<32>::new(token_id.to_bytes()),
+                tokenManagerType: token_manager_type.into(),
+                params: Vec::<u8>::from(params).into(),
+            }
```
Comment (cgorenflo) [inline packages/interchain-token-service-api/src/abi.rs:102]: seems to me that the mapping should be defined on `ItsMessage`,

File: packages/interchain-token-service-api/src/abi.rs [hunk 15]
```diff
@@ -0,0 +1,469 @@
...
+        /// ITS edge source contract -> ITS Hub GMP call: Set to the true destination chain name.
+        /// ITS Hub -> ITS edge destination contract: Set to the true source chain name.
+        string remote_chain;
+        bytes message;
+    }
+}
+
+impl ITSRoutedMessage {
+    pub fn abi_encode(&self) -> HexBinary {
+        let message: Vec<u8> = match self.message.clone() {
```
Comment (cgorenflo) [inline packages/interchain-token-service-api/src/abi.rs:62]: why is this defined on a reference only to then clone the message?

**Review by cgorenflo — CHANGES_REQUESTED** [review]

**Review by Foivos — APPROVED** [review]

File: packages/interchain-token-service-api/src/abi.rs (hunk 9 above)
Comment (milapsheth) [reply packages/interchain-token-service-api/src/abi.rs]: Yeah, we can have two message types on the wire to distinguish it. The downside is that now you can have message types that are only valid in a certain direction (send to hub / receive from hub). Another option is to keep it the same on the wire but use send/receive type Rust types. But this makes decoding context sensitive which seems confusing. Discussed with SC team, and we decided to go with two separate message types

File: packages/interchain-token-service-api/src/testdata/deploy_interchain_token_encode_decode.golden
[the remark this answers is not in the document]
Comment (milapsheth) [reply packages/interchain-token-service-api/src/testdata/deploy_interchain_token_encode_decode.golden:1]: yeah, don't see a way to force a newline. Since this in a test file, I'm ok to not add it. Otherwise we can look at adding a script that does it post generation

File: packages/interchain-token-service-api/src/abi.rs (hunk 13 above)
Comment (milapsheth) [reply packages/interchain-token-service-api/src/abi.rs]: We might have other encoding types in the future potentially. This module is responsible for only adding ABI functionality

File: packages/interchain-token-service-api/src/primitives.rs (hunk 12 above)
Comment (milapsheth) [reply packages/interchain-token-service-api/src/primitives.rs]: https://github.com/axelarnetwork/axelar-amplifier/pull/482#discussion_r1686782811

File: packages/interchain-token-service-api/src/primitives.rs [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+impl TokenId {
+    #[inline]
+    pub const fn new(id: [u8; 32]) -> Self {
+        Self(id)
+    }
+
+    #[inline]
+    pub const fn to_bytes(&self) -> [u8; 32] {
+        self.0
+    }
```
[the remark this answers is not in the document]
Comment (milapsheth) [reply packages/interchain-token-service-api/src/primitives.rs]: Converted to `into`. The compiler will inline it so it shouldn't copy. To ensure that the inlining optimization works [cross-crate boundaries](https://matklad.github.io/2021/07/09/inline-in-rust.html), I added the `#[inline]` macro. I've seen this as a common pattern for other crates.

File: interchain-token-service/src/error.rs [hunk 17]
```diff
@@ -0,0 +1,10 @@
+use axelar_wasm_std_derive::IntoContractError;
+use thiserror::Error;
+
+#[derive(Error, Debug, PartialEq, IntoContractError)]
+pub enum Error {
+    #[error("failed to decode ITS message: {0}")]
+    InvalidMessage(String),
```
Comment (cgorenflo) [inline interchain-token-service/src/error.rs:7]: I'm not a fan of how the error conversions are handled. This is exactly what error-stack is for:
define it as 
```
#[error("failed to decode ITS message")]
```

and then use
```
result.change_context(Error::InvalidMessage)
```
instead of 
```
result.map_err(|e| Report::new(Error::InvalidMessage(e.to_string())));
```

**Review by cgorenflo — APPROVED** [review]
